File: archive/OnlyRequiredFiles/Eq2ImgSettings.cs
using System;
using System.Web;
using System.Configuration;
using System.Globalization;
using System.Web.Caching;

namespace Astrila.Eq2Img
{
	/// <summary>
	/// Summary description for ShowEqSettings.
	/// </summary>
	[Serializable]
	public class Eq2ImgSettings
	{
		//Defines the default values for the settings if they weren't specified in 
		//web.config
		private const int DefaultClientCacheAgeInDays = 180;
		private const string DefaultCacheFolder = @"CachedEqImages";
		private const bool DefaultIsCacheFolderRelative = true;
		private const MimeTexUsageType DefaultMimeTexUsage = MimeTexUsageType.InProc;
		private const string DefaultMimeTexExeRelativeFilePath = @"cgi\MimeTex.exe";
		private const string DefaultMimeTexFullCgiUrl = @"cgi/MimeTex.exe";
		private const string DefaultMimeTexExternalServerUrl = @"http://www.ShitalShah.com/?{0}";
		//This may ease DoS attack but its required for figure type images that MimeTeX supports
		private const int DefaultEquationMaxSize = 10000;

		//Set up the private members for the properties
		private TimeSpan _ClientCacheAge =  TimeSpan.FromDays(DefaultClientCacheAgeInDays);
		private string _CacheFolder = DefaultCacheFolder;
		private bool _IsCacheFolderRelative = DefaultIsCacheFolderRelative;
		private MimeTexUsageType _MimeTexUsage = DefaultMimeTexUsage;
		private string _MimeTexExeRelativeFilePath = DefaultMimeTexExeRelativeFilePath;
		private string _MimeTexFullCgiUrl = DefaultMimeTexFullCgiUrl;
		private string _MimeTexExternalServerUrl = DefaultMimeTexExternalServerUrl;
		private int _EquationMaxSize = DefaultEquationMaxSize;
		private string _AdminKey = string.Empty;

		//By default the cache files created for equations would have this time stamp
		private DateTime _LastModifiedDateHttpHeader = DateTime.Now;

#region "Property Getters/Setters
		public string AdminKey
		{
			get {return _AdminKey;}
			set {_AdminKey = value;}
		}
		public TimeSpan ClientCacheAge
		{
			get {return _ClientCacheAge;}
			set {_ClientCacheAge = value;}
		}
		public int EquationMaxSize
		{
			get {return _EquationMaxSize;}
			set {_EquationMaxSize = value;}
		}
		public DateTime ClientCacheExpiration
		{
			get {return DateTime.Now.Add(_ClientCacheAge);}
		}
		public DateTime LastModifiedDateHttpHeader
		{
			get {return _LastModifiedDateHttpHeader;}
			set {_LastModifiedDateHttpHeader = value;}
		}
		public string CacheFolder
		{
			get {return _CacheFolder;}
			set {_CacheFolder = value;}
		}
		public bool IsCacheFolderRelative
		{
			get {return _IsCacheFolderRelative;}
			set {_IsCacheFolderRelative = value;}
		}
		public MimeTexUsageType MimeTexUsage
		{
			get {return _MimeTexUsage;}
			set {_MimeTexUsage = value;}
		}
		public string MimeTexExeRelativeFilePath
		{
			get {return _MimeTexExeRelativeFilePath;}
			set {_MimeTexExeRelativeFilePath = value;}
		}
		public string MimeTexFullCgiUrl
		{
			get {return _MimeTexFullCgiUrl;}
			set {_MimeTexFullCgiUrl = value;}
		}
		public string MimeTexExternalServerUrl
		{
			get {return _MimeTexExternalServerUrl;}
			set {_MimeTexExternalServerUrl = value;}
		}
		//TODO: late additions. Need to add code for these properties!
		public TimeSpan ClientCacheAgeForErrorMessages
		{
			get {return TimeSpan.FromMinutes(15);}
		}
		public DateTime ClientCacheExpirationForErrorMessages
		{
			get {return DateTime.Now.AddMinutes(15);}
		}
#endregion

		//Each setting name in the config file will have this prefix
		private const string ConfigPrefix = "Eq2Img_";	

		//Load settings from config file
		public void LoadFromConfiguration()
		{
			string adminKeyFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "AdminKey"];
			if (adminKeyFromConfig != null)
				_AdminKey = adminKeyFromConfig;
			else
				_AdminKey = string.Empty;

			string clientCacheAgeInSecondsFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "ClientCacheAgeInSeconds"];
			if (clientCacheAgeInSecondsFromConfig != null)
				_ClientCacheAge = TimeSpan.FromSeconds(int.Parse(clientCacheAgeInSecondsFromConfig, CultureInfo.InvariantCulture));
			else
				_ClientCacheAge = TimeSpan.FromDays(DefaultClientCacheAgeInDays);

			string isCacheFolderRelativeFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "IsCacheFolderRelative"];
			if (isCacheFolderRelativeFromConfig != null)
				_IsCacheFolderRelative = Convert.ToBoolean(isCacheFolderRelativeFromConfig, CultureInfo.InvariantCulture);
			else
				_IsCacheFolderRelative = DefaultIsCacheFolderRelative;

			string equationMaxSizeFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "EquationMaxSize"];
			if (equationMaxSizeFromConfig != null)
				_EquationMaxSize = Convert.ToInt32(equationMaxSizeFromConfig, CultureInfo.InvariantCulture);
			else
				_EquationMaxSize = DefaultEquationMaxSize;

			//Get the path where to store cache images. This coule be either relative to current web app path OR
			//absolute path on server
			string cacheFolderFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "ImageCacheFolder"];
			if (cacheFolderFromConfig != null)
			{
				if (_IsCacheFolderRelative)	//then build absolute path
					_CacheFolder = HttpContext.Current.Server.MapPath(cacheFolderFromConfig);
				else	//get absolute path as is
					_CacheFolder = cacheFolderFromConfig;
			}
			else
			{
				if (!_IsCacheFolderRelative)
				{
					//absolute path must be set if _IsCacheFolderRelative is set to false
					throw new ArgumentException(@"The setting in web.config '" + ConfigPrefix + "IsCacheFolderRelative" + "' is set to false but no absolute path has been provided", ConfigPrefix + "ImageCacheFolder");
				}
				else
					_CacheFolder = HttpContext.Current.Server.MapPath(DefaultCacheFolder);
			}

			string MimeTexUsageFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "MimeTexUsage"];
			if (MimeTexUsageFromConfig != null)
				_MimeTexUsage =(MimeTexUsageType) System.Enum.Parse(typeof(MimeTexUsageType), MimeTexUsageFromConfig, true);
			else
				_MimeTexUsage = DefaultMimeTexUsage;

			string MimeTexExeRelativeFilePathFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "ExeRelativeFilePath"];
			if (MimeTexExeRelativeFilePathFromConfig != null)
				_MimeTexExeRelativeFilePath = MimeTexExeRelativeFilePathFromConfig;
			else
				_MimeTexExeRelativeFilePath = DefaultMimeTexExeRelativeFilePath;

			string MimeTexFullCgiUrlFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "FullCGIUrl"];
			if (MimeTexFullCgiUrlFromConfig != null)
				_MimeTexFullCgiUrl = MimeTexFullCgiUrlFromConfig;
			else
				_MimeTexFullCgiUrl = DefaultMimeTexFullCgiUrl;

			string MimeTexExternalServerUrlFromConfig = ConfigurationSettings.AppSettings[ConfigPrefix + "ExternalServerUrl"];
			if (MimeTexExternalServerUrlFromConfig != null)
				_MimeTexExternalServerUrl = MimeTexExternalServerUrlFromConfig;
			else
				_MimeTexExternalServerUrl = DefaultMimeTexExternalServerUrl;

			_LastModifiedDateHttpHeader = DateTime.Now;
		}

		//Gets the currenly cached settings object OR loads new from config
		private static object _CachedSettingsLocker = new object();
		public static Eq2ImgSettings GetFromContext(HttpContext context)
		{
			if (context == null) throw new ArgumentNullException("context", "The context parameter can not be null");

			Eq2ImgSettings settings = (Eq2ImgSettings) context.Cache["Eq2ImgSettings"];
			if (settings == null)
			{
				lock(_CachedSettingsLocker)
				{
					settings = (Eq2ImgSettings) context.Cache["Eq2ImgSettings"];
					if (settings == null)
					{
						settings = new Eq2ImgSettings();
						settings.LoadFromConfiguration();
						context.Cache.Add("Eq2ImgSettings", settings, new CacheDependency(context.Server.MapPath("web.config")), DateTime.MaxValue, TimeSpan.Zero, CacheItemPriority.Low, null);
					}
				}
			}
			else {/* return existing settings */}; 

			return settings;
		}
	}

	public enum MimeTexUsageType
	{
		None = 0,
		InProc = 1,
		OutProc = 2,
		Cgi = 3,
		ExternalUrl = 4
	}
}

File: archive/VB.Net Version/Eq2ImgVB/Eq2ImgSettings.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Web
Imports System.Configuration
Imports System.Globalization
Imports System.Web.Caching

Namespace Astrila.Eq2Img
	''' <summary>
	''' Summary description for ShowEqSettings.
	''' </summary>
	<Serializable> _
	Public Class Eq2ImgSettings
		'Defines the default values for the settings if they weren't specified in 
		'web.config
		Private Const DefaultClientCacheAgeInDays As Integer = 180
		Private Const DefaultCacheFolder As String = "CachedEqImages"
		Private Const DefaultIsCacheFolderRelative As Boolean = True
		Private Const DefaultMimeTexUsage As MimeTexUsageType = MimeTexUsageType.InProc
		Private Const DefaultMimeTexExeRelativeFilePath As String = "cgi\MimeTex.exe"
		Private Const DefaultMimeTexFullCgiUrl As String = "cgi/MimeTex.exe"
        Private Const DefaultMimeTexExternalServerUrl As String = "http://www.ShitalShah.com/?{0}"
        'This may ease DoS attack but its required for figure type images that MimeTeX supports
        Private Const DefaultEquationMaxSize As Integer = 10000

		'Set up the private members for the properties
		Private _ClientCacheAge As TimeSpan = TimeSpan.FromDays(DefaultClientCacheAgeInDays)
		Private _CacheFolder As String = DefaultCacheFolder
		Private _IsCacheFolderRelative As Boolean = DefaultIsCacheFolderRelative
		Private _MimeTexUsage As MimeTexUsageType = DefaultMimeTexUsage
		Private _MimeTexExeRelativeFilePath As String = DefaultMimeTexExeRelativeFilePath
		Private _MimeTexFullCgiUrl As String = DefaultMimeTexFullCgiUrl
		Private _MimeTexExternalServerUrl As String = DefaultMimeTexExternalServerUrl
		Private _EquationMaxSize As Integer = DefaultEquationMaxSize
		Private _AdminKey As String = String.Empty

		'By default the cache files created for equations would have this time stamp
		Private _LastModifiedDateHttpHeader As DateTime = DateTime.Now

#region "Property Getters/Setters
		Public Property AdminKey() As String
			Get
				Return _AdminKey
			End Get
			Set
				_AdminKey = Value
			End Set
		End Property
		Public Property ClientCacheAge() As TimeSpan
			Get
				Return _ClientCacheAge
			End Get
			Set
				_ClientCacheAge = Value
			End Set
		End Property
		Public Property EquationMaxSize() As Integer
			Get
				Return _EquationMaxSize
			End Get
			Set
				_EquationMaxSize = Value
			End Set
		End Property
		Public ReadOnly Property ClientCacheExpiration() As DateTime
			Get
				Return DateTime.Now.Add(_ClientCacheAge)
			End Get
		End Property
		Public Property LastModifiedDateHttpHeader() As DateTime
			Get
				Return _LastModifiedDateHttpHeader
			End Get
			Set
				_LastModifiedDateHttpHeader = Value
			End Set
		End Property
		Public Property CacheFolder() As String
			Get
				Return _CacheFolder
			End Get
			Set
				_CacheFolder = Value
			End Set
		End Property
		Public Property IsCacheFolderRelative() As Boolean
			Get
				Return _IsCacheFolderRelative
			End Get
			Set
				_IsCacheFolderRelative = Value
			End Set
		End Property
		Public Property MimeTexUsage() As MimeTexUsageType
			Get
				Return _MimeTexUsage
			End Get
			Set
				_MimeTexUsage = Value
			End Set
		End Property
		Public Property MimeTexExeRelativeFilePath() As String
			Get
				Return _MimeTexExeRelativeFilePath
			End Get
			Set
				_MimeTexExeRelativeFilePath = Value
			End Set
		End Property
		Public Property MimeTexFullCgiUrl() As String
			Get
				Return _MimeTexFullCgiUrl
			End Get
			Set
				_MimeTexFullCgiUrl = Value
			End Set
		End Property
		Public Property MimeTexExternalServerUrl() As String
			Get
				Return _MimeTexExternalServerUrl
			End Get
			Set
				_MimeTexExternalServerUrl = Value
			End Set
		End Property
		'TODO: late additions. Need to add code for these properties!
		Public ReadOnly Property ClientCacheAgeForErrorMessages() As TimeSpan
			Get
				Return TimeSpan.FromMinutes(15)
			End Get
		End Property
		Public ReadOnly Property ClientCacheExpirationForErrorMessages() As DateTime
			Get
				Return DateTime.Now.AddMinutes(15)
			End Get
		End Property
#End Region

		'Each setting name in the config file will have this prefix
		Private Const ConfigPrefix As String = "Eq2Img_"

		'Load settings from config file
		Public Sub LoadFromConfiguration()
			Dim adminKeyFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "AdminKey")
			If Not adminKeyFromConfig Is Nothing Then
				_AdminKey = adminKeyFromConfig
			Else
				_AdminKey = String.Empty
			End If

			Dim clientCacheAgeInSecondsFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "ClientCacheAgeInSeconds")
			If Not clientCacheAgeInSecondsFromConfig Is Nothing Then
				_ClientCacheAge = TimeSpan.FromSeconds(Integer.Parse(clientCacheAgeInSecondsFromConfig, CultureInfo.InvariantCulture))
			Else
				_ClientCacheAge = TimeSpan.FromDays(DefaultClientCacheAgeInDays)
			End If

			Dim isCacheFolderRelativeFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "IsCacheFolderRelative")
			If Not isCacheFolderRelativeFromConfig Is Nothing Then
				_IsCacheFolderRelative = Convert.ToBoolean(isCacheFolderRelativeFromConfig, CultureInfo.InvariantCulture)
			Else
				_IsCacheFolderRelative = DefaultIsCacheFolderRelative
			End If

			Dim equationMaxSizeFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "EquationMaxSize")
			If Not equationMaxSizeFromConfig Is Nothing Then
				_EquationMaxSize = Convert.ToInt32(equationMaxSizeFromConfig, CultureInfo.InvariantCulture)
			Else
				_EquationMaxSize = DefaultEquationMaxSize
			End If

			'Get the path where to store cache images. This coule be either relative to current web app path OR
			'absolute path on server
			Dim cacheFolderFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "ImageCacheFolder")
			If Not cacheFolderFromConfig Is Nothing Then
				If _IsCacheFolderRelative Then 'then build absolute path
					_CacheFolder = HttpContext.Current.Server.MapPath(cacheFolderFromConfig)
				Else 'get absolute path as is
					_CacheFolder = cacheFolderFromConfig
				End If
			Else
				If (Not _IsCacheFolderRelative) Then
					'absolute path must be set if _IsCacheFolderRelative is set to false
					Throw New ArgumentException("The setting in web.config '" & ConfigPrefix & "IsCacheFolderRelative" & "' is set to false but no absolute path has been provided", ConfigPrefix & "ImageCacheFolder")
				Else
					_CacheFolder = HttpContext.Current.Server.MapPath(DefaultCacheFolder)
				End If
			End If

			Dim MimeTexUsageFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "MimeTexUsage")
			If Not MimeTexUsageFromConfig Is Nothing Then
				_MimeTexUsage =CType(System.Enum.Parse(GetType(MimeTexUsageType), MimeTexUsageFromConfig, True), MimeTexUsageType)
			Else
				_MimeTexUsage = DefaultMimeTexUsage
			End If

			Dim MimeTexExeRelativeFilePathFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "ExeRelativeFilePath")
			If Not MimeTexExeRelativeFilePathFromConfig Is Nothing Then
				_MimeTexExeRelativeFilePath = MimeTexExeRelativeFilePathFromConfig
			Else
				_MimeTexExeRelativeFilePath = DefaultMimeTexExeRelativeFilePath
			End If

			Dim MimeTexFullCgiUrlFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "FullCGIUrl")
			If Not MimeTexFullCgiUrlFromConfig Is Nothing Then
				_MimeTexFullCgiUrl = MimeTexFullCgiUrlFromConfig
			Else
				_MimeTexFullCgiUrl = DefaultMimeTexFullCgiUrl
			End If

			Dim MimeTexExternalServerUrlFromConfig As String = ConfigurationSettings.AppSettings(ConfigPrefix & "ExternalServerUrl")
			If Not MimeTexExternalServerUrlFromConfig Is Nothing Then
				_MimeTexExternalServerUrl = MimeTexExternalServerUrlFromConfig
			Else
				_MimeTexExternalServerUrl = DefaultMimeTexExternalServerUrl
			End If

			_LastModifiedDateHttpHeader = DateTime.Now
		End Sub

		'Gets the currenly cached settings object OR loads new from config
		Private Shared _CachedSettingsLocker As Object = New Object()
		Public Shared Function GetFromContext(ByVal context As HttpContext) As Eq2ImgSettings
			If context Is Nothing Then
                Throw New ArgumentNullException("context", "The context parameter can not be null")
			End If

			Dim settings As Eq2ImgSettings = CType(context.Cache("Eq2ImgSettings"), Eq2ImgSettings)
			If settings Is Nothing Then
				SyncLock _CachedSettingsLocker
					settings = CType(context.Cache("Eq2ImgSettings"), Eq2ImgSettings)
					If settings Is Nothing Then
						settings = New Eq2ImgSettings()
						settings.LoadFromConfiguration()
						context.Cache.Add("Eq2ImgSettings", settings, New CacheDependency(context.Server.MapPath("web.config")), DateTime.MaxValue, TimeSpan.Zero, CacheItemPriority.Low, Nothing)
					End If
				End SyncLock
            Else
                'return existing settings
            End If

            Return settings
		End Function
	End Class

	Public Enum MimeTexUsageType
		None = 0
		InProc = 1
		OutProc = 2
		Cgi = 3
		ExternalUrl = 4
	End Enum
End Namespace
